### PR TITLE
feat: precision prop and amp support for lightning adapter [DET-5116]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- add parameters `precision`, `amp_level`, and `amp_backend` to LightningAdapter to allow user to set up automatic mixed precision using a similar API to trainer and without calling `auto_amp` or `configure_apex_amp`.
- set precision property

## Test Plan
gpu e2e tests are added and passed here :
https://app.circleci.com/pipelines/github/determined-ai/determined?branch=ptl-amp

## Commentary (optional)

users can opt to avoid using the added parameters and set up AMP themselves using PyTorchTrial APIs

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/pytorch/lightning/_adapter.py [hunk 1]
```diff
@@ -142,6 +150,17 @@ def __init__(self, context: PyTorchTrialContext) -> None:
                 lm = mnist.LightningMNISTClassifier(lr=context.get_hparam('learning_rate'))
                 super().__init__(context, lightning_module=lm)
 
+        Arguments:
+            context (``PyTorchTrialContext``)
```
Comment (shiyuann) [inline harness/determined/pytorch/lightning/_adapter.py:154]: Is it some convention to use `(` and `)` for both types and values to choose?

File: harness/determined/pytorch/lightning/_adapter.py [hunk 2]
```diff
@@ -152,16 +171,37 @@ def __init__(self, context: PyTorchTrialContext) -> None:
         optimizers, _ = self.setup_optimizers_schedulers()
         pls.optimizers = optimizers
 
+        # CHECK our `check` usage vs raising exceptions seems inconsistent.
```
Comment (shiyuann) [inline harness/determined/pytorch/lightning/_adapter.py:174]: yes. I'd say it's fine to use either of them until we decide to clean them up

File: harness/determined/pytorch/lightning/_adapter.py [hunk 3]
```diff
@@ -152,16 +171,37 @@ def __init__(self, context: PyTorchTrialContext) -> None:
...
         # set lightning_module properties
         pls.lm.use_ddp = False
         pls.lm.use_ddp2 = False
         pls.lm.use_dp = False
         pls.lm.use_tpu = False
         type(pls.lm).local_rank = context.distributed.get_local_rank()  # type: ignore
         type(pls.lm).global_rank = context.distributed.get_rank()  # type: ignore
-        pls.lm.use_amp = context.experimental._auto_amp or context._use_apex
+        pls.lm.use_amp = use_amp
+        pls.lm.precision = "mixed" if use_amp else precision  # type: ignore
```
Comment (shiyuann) [inline harness/determined/pytorch/lightning/_adapter.py:187]: Does this have the same behavior as `LightningModule.precision`?

File: harness/determined/pytorch/lightning/_adapter.py [hunk 4]
```diff
@@ -152,16 +171,37 @@ def __init__(self, context: PyTorchTrialContext) -> None:
         optimizers, _ = self.setup_optimizers_schedulers()
         pls.optimizers = optimizers
 
+        # CHECK our `check` usage vs raising exceptions seems inconsistent.
+        check.check_in(precision, {16, 32}, "only precisions 16 & 32 are supported.")
+
+        use_amp = context.experimental._auto_amp or context._use_apex
```
Comment (shiyuann) [inline harness/determined/pytorch/lightning/_adapter.py:177]: This seems wrong. Because `_use_apex` is set in `configure_apex_amp` and the same for `_auto_amp`

**Comment by shiyuann** [comment]
@hamidzr can you also add e2e test for this?

File: harness/determined/pytorch/lightning/_adapter.py (hunk 3 above)
Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:187]: yeah if I think so. through `accelerator.precision`. I'll share the reference

File: harness/determined/pytorch/lightning/_adapter.py (hunk 1 above)
Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:154]: I think of the constant values as union literal types. python type would shape it like `Union[Literal['x'], Literal['y']]`

Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:154]: what do you suggest

File: harness/determined/pytorch/lightning/_adapter.py (hunk 4 above)
Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:177]: I'm trying to set use_amp without relying on the precision property 🤔 let me see

File: harness/determined/pytorch/lightning/_adapter.py (hunk 1 above)
Comment (shiyuann) [reply harness/determined/pytorch/lightning/_adapter.py:154]: we follow [this convention](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html) since the docstring will be used for the sphinx doc.

Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:154]: I had just followed the pattern from pytorch_trial. this is helpful thanks. I'll update it

File: harness/determined/pytorch/lightning/_adapter.py (hunk 3 above)
Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:187]: AFAICT comes from `class MixedPrecisionPlugin(PrecisionPlugin):` base class

File: harness/determined/pytorch/lightning/_adapter.py (hunk 1 above)
Comment (shiyuann) [reply harness/determined/pytorch/lightning/_adapter.py]: Sorry about the confusion! I actually meant this:

```
opt_level (str, optional, default="O1"):  Pure or mixed precision optimization level.
                Accepted values are "O0", "O1", "O2", and "O3", explained in detail above.
```
This also works with sphinx docstring. But I sent a wrong link.

**Comment by hamidzr** [comment]
one of the github check's are stuck. I'll rebase to retrigger it.

File: e2e_tests/.flake8 [hunk 5]
```diff
@@ -36,3 +36,9 @@ import-order-style = edited
 # flake8-quotes
 inline-quotes = "
 multiline-quotes = """
+
```
Comment (hamidzr) [inline e2e_tests/.flake8:39]: https://github.com/determined-ai/determined/pull/2140 should remove the need for this change but in the meantime we can do something like this

File: e2e_tests/tests/fixtures/pytorch_lightning_amp/apex_amp_model_def.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+"""
+This example demonstrates how to modify a model to use PyTorch's support for
```
Comment (hamidzr) [inline e2e_tests/tests/fixtures/pytorch_lightning_amp/apex_amp_model_def.py:2]: pytorch lightning.

File: e2e_tests/tests/fixtures/pytorch_lightning_amp/auto_amp_model_def.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+"""
+This example demonstrates how to modify a model to use PyTorch's native AMP
```
Comment (hamidzr) [inline e2e_tests/tests/fixtures/pytorch_lightning_amp/auto_amp_model_def.py:2]: update the comment for lightning

**Review by shiyuann — APPROVED** [review]

File: harness/determined/pytorch/lightning/_adapter.py [hunk 8]
```diff
@@ -120,7 +121,14 @@ class LightningAdapter(PyTorchTrial):
...
     """
 
-    def __init__(self, context: PyTorchTrialContext, lightning_module: pl.LightningModule):
+    def __init__(
+        self,
+        context: PyTorchTrialContext,
+        lightning_module: pl.LightningModule,
+        precision: Union[Literal[32], Literal[16]] = 32,
+        amp_backend: Union[Literal["native"], Literal["apex"]] = "native",
+        amp_level: Union[Literal["O0", "O1", "O2", "O3"]] = "O2",
```
Comment (shiyuann) [inline harness/determined/pytorch/lightning/_adapter.py:130]: Should this be `O0`?

Comment (hamidzr) [reply harness/determined/pytorch/lightning/_adapter.py:130]: it's visually confusing but it is a capital O + the number 0 from what I can tell. I confirmed by searching for the expected value.
